lib/iris/fileformats/grib/__init__.py

```diff
@@ -179,12 +177,14 @@ def __init__(self, grib_message, grib_fh=None):
             # The byte offset requires to be reset back to the first byte
             # of this message. The file pointer offset is always at the end
             # of the current message due to the grib-api reading the message.
-            proxy = GribDataProxy(shape, np.zeros(0).dtype, np.nan,
-                                  grib_fh.name,
+            proxy = GribDataProxy(shape, np.zeros(0).dtype, grib_fh.name,
                                   offset - message_length)
             self._data = as_lazy_data(proxy)
         else:
-            self.data = _message_values(grib_message, shape)
+            values_array = _message_values(grib_message, shape)
+            # mask where the values are nan
+            self.data = convert_nans_array(values_array,
+                                           nans_replacement=ma.masked)
```

Comment:
Yep, previously `_message_value` was always returning a masked array but now it's returning a nan array which we need to convert to a masked array here, if it's not deferred. If it is deferred we hand it off to dask which will eventually turn it into a masked array.